src/shacl/tokens.ts

```diff
@@ -0,0 +1,347 @@
+const { turtleTokenTypes, turtleTokenMap } = require('../turtle/tokens');
+const { sparqlTokenMap } = require('../sparql/tokens');
+import { TokenType, createToken, Lexer } from 'chevrotain';
+import memoize from 'memoize-one';
+import { getAsTypedTuple } from 'helpers/types';
+import isDeepEqual from 'lodash.isequal';
+
+const shaclIriNamespace = 'http://www.w3.org/ns/shacl#';
+const xsdIriNamespace = 'http://www.w3.org/2001/XMLSchema#';
+
+// Token categories, useful for making the parser rules simpler.
+export const categoryTokenMap = {
+  IriTakingPredicate: createToken({
+    name: 'IriTakingPredicate',
+    pattern: Lexer.NA,
+  }),
+  NodeKindIRI: createToken({
+    name: 'NodeKindIRI',
+    pattern: Lexer.NA,
+  }),
+  IntTakingPredicate: createToken({
+    name: 'IntTakingPredicate',
+    pattern: Lexer.NA,
+  }),
+  StringLiteralQuoteTakingPredicate: createToken({
+    name: 'StringLiteralQuoteTakingPredicate',
+    pattern: Lexer.NA,
+  }),
+  LangStringTakingPredicate: createToken({
+    name: 'LangStringTakingPredicate',
+    pattern: Lexer.NA,
+  }),
+  BooleanTakingPredicate: createToken({
+    name: 'BooleanTakingPredicate',
+    pattern: Lexer.NA,
+  }),
+  ShapeExpectingPredicate: createToken({
+    name: 'ShapeExpectingPredicate',
+    pattern: Lexer.NA,
+  }),
+  AnyLiteralTakingPredicate: createToken({
+    name: 'AnyLiteralTakingPredicate',
+    pattern: Lexer.NA,
+  }),
+};
+
+export const categoryTokens = Object.keys(categoryTokenMap).map(
+  (key) => categoryTokenMap[key]
+);
+
+const localNamesByCategory = {
+  IriTakingPredicate: getAsTypedTuple(
+    'class',
+    'datatype',
+    'equals',
+    'disjoint',
+    'lessThan',
+    'lessThanOrEquals',
+    'targetClass',
+    'targetSubjectsOf',
+    'targetObjectsOf',
+    'severity'
+  ),
+  NodeKindIRI: getAsTypedTuple(
+    'IRI',
+    'BlankNode',
+    'Literal',
+    'BlankNodeOrIRI',
+    'BlankNodeOrLiteral',
+    'IRIOrLiteral'
+  ),
+  IntTakingPredicate: getAsTypedTuple(
+    'minCount',
+    'maxCount',
+    'minLength',
+    'maxLength',
+    'qualifiedMinCount',
+    'qualifiedMaxCount'
+  ),
+  StringLiteralQuoteTakingPredicate: getAsTypedTuple(
+    'pattern',
+    'flags',
+    'prefix',
+    'namespace'
+  ),
+  LangStringTakingPredicate: getAsTypedTuple('message', 'labelTemplate'),
+  BooleanTakingPredicate: getAsTypedTuple(
+    'uniqueLang',
+    'qualifiedValueShapesDisjoint',
+    'closed',
+    'deactivated',
+    'optional'
+  ),
+  ShapeExpectingPredicate: getAsTypedTuple(
+    'not',
+    'node',
+    'property',
+    'qualifiedValueShape',
+    'sparql',
+    'declare',
+    'prefixes',
+    'parameter',
+    'nodeValidator',
+    'propertyValidator',
+    'validator'
+  ),
+  AnyLiteralTakingPredicate: getAsTypedTuple(
```

Comment:
I see this in the spec..but for my understanding, are these any literal b/c `<` and `>` can be used for any literal type?

Reply:
Probably so. Also because, for these, the spec says, "The values . . . are literals," whereas, for other predicates (e.g., "sh:uniqueLang"), it specifies the types, as in "The values of sh:uniqueLang in a shape are literals **with datatype xsd:boolean.**" I didn't want to be any more restrictive.

Reply:
makes sense